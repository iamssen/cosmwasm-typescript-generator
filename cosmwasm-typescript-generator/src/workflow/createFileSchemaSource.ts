import * as fs from 'fs';
import * as glob from 'glob';
import * as path from 'path';
import { SchemaSource } from '../types';

export async function createFileSchemaSource(
  schemaRoot: string,
): Promise<SchemaSource> {
  const contractNames = fs
    .readdirSync(schemaRoot)
    .filter((contract) =>
      fs.statSync(path.resolve(schemaRoot, contract)).isDirectory(),
    );

  return contractNames.map((contract) => {
    return {
      contract,
      schema: async () => {
        const schemaFiles = glob.sync(`${schemaRoot}/${contract}/*.json`);
        const sources = await Promise.all(
          schemaFiles.map((schemaFile) =>
            fs.readFileSync(schemaFile, { encoding: 'utf8' }),
          ),
        );
        return sources.map((jsonSource) => JSON.parse(jsonSource));
      },
    };
  });
}
