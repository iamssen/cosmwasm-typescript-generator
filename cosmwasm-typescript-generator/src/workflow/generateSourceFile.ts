import { camelCase } from 'change-case';
import {
  createSourceFile,
  factory,
  InterfaceDeclaration,
  ModuleDeclaration,
  NodeFlags,
  ScriptTarget,
  SourceFile,
  SyntaxKind,
} from 'typescript';
import { createTypeStore } from '../transform/createTypeStore';
import { parseGroupSchema } from '../transform/parseGroupSchema';
import { parseSingleSchema } from '../transform/parseSingleSchema';
import { SchemaSource } from '../types';

export async function generateSourceFile(
  source: SchemaSource,
): Promise<SourceFile> {
  const contractDeclarations: ModuleDeclaration[] = [];
  const { registerDefinitions, getTypeNode, getTypeStatements } =
    createTypeStore();

  for (const { contract: contractName, schema: getSchemas } of source) {
    const statements: (ModuleDeclaration | InterfaceDeclaration)[] = [];
    const schemas = await getSchemas();

    for (const schema of schemas) {
      if (schema.definitions) {
        registerDefinitions(schema.definitions);
      }

      try {
        if (
          schema.title &&
          (Array.isArray(schema.anyOf) || Array.isArray(schema.oneOf))
        ) {
          statements.push(parseGroupSchema(schema, getTypeNode));
        } else if (schema.title) {
          statements.push(parseSingleSchema(schema, getTypeNode));
        }
      } catch (error) {
        console.error(
          `Failed to parse ${contractName}/#${schema.title}`,
          error,
        );
      }
    }

    const contractDeclaration = factory.createModuleDeclaration(
      undefined,
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      factory.createIdentifier(camelCase(contractName)),
      factory.createModuleBlock(statements),
      NodeFlags.Namespace,
    );

    contractDeclarations.push(contractDeclaration);
  }

  return factory.updateSourceFile(
    createSourceFile('source.ts', '', ScriptTarget.ESNext),
    [...getTypeStatements(), ...contractDeclarations],
  );
}
