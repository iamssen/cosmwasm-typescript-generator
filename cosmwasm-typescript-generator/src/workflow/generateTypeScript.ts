import { createPrinter, SourceFile } from 'typescript';

export function generateTypeScript(source: SourceFile): string {
  const printer = createPrinter();
  return [
    `// DO NOT EDIT MANUALLY : THIS FILE IS AUTO CREATED`,
    `/* eslint-disable */`,
    `// prettier-ignore`,
    `// noinspection JSUnusedGlobalSymbols`,
    ``,
    printer.printFile(source).replace(/\/\*/g, '/**'),
  ].join('\n');
}