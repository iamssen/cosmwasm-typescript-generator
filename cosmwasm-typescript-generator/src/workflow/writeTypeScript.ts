import fs from 'fs';
import { SourceFile } from 'typescript';
import { generateTypeScript } from './generateTypeScript';

export function writeTypeScript(source: SourceFile, outFile: string) {
  const code = generateTypeScript(source);

  fs.writeFileSync(outFile, code, {
    encoding: 'utf8',
  });
}
