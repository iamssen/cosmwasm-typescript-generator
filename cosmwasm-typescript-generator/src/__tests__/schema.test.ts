import { generateTypeScript } from '../workflow/generateTypeScript';
import { generateSourceFile } from '../workflow/generateSourceFile';
import { createFileSchemaSource } from '../workflow/createFileSchemaSource';
import * as path from 'path';

describe('schema test', () => {
  test('should parse counter schema', async () => {
    const schemaSource = await createFileSchemaSource(
      path.resolve(__dirname, '../../__fixtures__/schema/counter'),
    );
    
    const sourceFile = await generateSourceFile(schemaSource);
    
    const code = generateTypeScript(sourceFile);
    
    expect(code).toMatchSnapshot();
  });
});
