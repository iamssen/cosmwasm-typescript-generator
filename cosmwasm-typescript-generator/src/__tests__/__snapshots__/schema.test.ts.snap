// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema test should parse counter schema 1`] = `
"// DO NOT EDIT MANUALLY : THIS FILE IS AUTO CREATED
/* eslint-disable */
// prettier-ignore
// noinspection JSUnusedGlobalSymbols

export type int32 = number;
/**A human readable address.

In Cosmos, this is typically bech32 encoded. But for multi-chain smart contracts no assumptions should be made other than being UTF-8 encoded and of reasonable length.

This type represents a validated address. It can be created in the following ways 1. Use \`Addr::unchecked(input)\` 2. Use \`let checked: Addr = deps.api.addr_validate(input)?\` 3. Use \`let checked: Addr = deps.api.addr_humanize(canonical_addr)?\` 4. Deserialize from JSON. This must only be done from JSON that was validated before such as a contract's state. \`Addr\` must not be used in messages sent by the user because this would result in unvalidated instances.

This type is immutable. If you really need to mutate it (Really? Are you sure?), create a mutable copy using \`let mut mutable = Addr::to_string()\` and operate on that \`String\` instance.*/
export type Addr = string;
export namespace counter {
    /**Response type of QueryMsg.GetCount*/
    export interface CountResponse {
        /**count property*/
        count: int32;
    }
    export namespace ExecuteMsg {
        /**increment count
        
        # Example
        
        \`\`\`json { \\"increment\\": {} } \`\`\`*/
        export interface Increment {
            increment: {};
        }
        /**decrement count
        
        # Example
        
        \`\`\`json { \\"increment\\": {} } \`\`\`*/
        export interface Decrement {
            decrement: {};
        }
        /**reset count
        
        # Example
        
        \`\`\`json { \\"reset\\": { \\"count\\": 0 } } \`\`\`*/
        export interface Reset {
            reset: {
                /**count property!!!*/
                count: int32;
            };
        }
    }
    export interface InstantiateMsg {
        /**count property?*/
        count: int32;
    }
    export namespace QueryMsg {
        /**GetCount returns the current count as a json-encoded number*/
        export interface GetCount {
            get_count: {};
        }
    }
    export interface State {
        /**count state*/
        count: int32;
        /**contract owner*/
        owner: Addr;
    }
}
"
`;
